cpp/src/parquet/arrow/schema.cc

```diff
@@ -1009,6 +1008,20 @@ Result<bool> ApplyOriginalMetadata(const Field& origin_field, SchemaField* infer
       // Origin type is restored as Arrow should be considered the source of truth.
       inferred->field = inferred->field->WithType(origin_type);
       RETURN_NOT_OK(ApplyOriginalStorageMetadata(origin_field, inferred));
+    } else if (inferred_type->id() == ::arrow::Type::EXTENSION &&
+               ex_type.extension_name() == std::string("arrow.json") &&
+               ::arrow::extension::JsonExtensionType::IsSupportedStorageType(
+                   inferred_type->storage_id())) {
```

Comment:
This condition `IsSupportedStorageType(inferred_type->storage_id()` should always be true, right? We cannot infer a json type with an incorrect storage type.

Reply:
Indeed. Removing this check.